File: python/luhn/luhn.py
import string

class Luhn(object):
  def __init__(self, number):
    self._numbers = [ch for ch in number if not ch.isspace()]
    self._all_digits = all(c in string.digits for c in self._numbers)

  def is_valid(self):
    if not (len(self._numbers) > 1 and self._all_digits):
      return False

    digits = [int(digit) for digit in self._numbers]
  
    for i in range(len(digits)-2, -1, -2):
      digits[i] = digits[i] * 2

      if digits[i] > 9:
        digits[i] = digits[i] - 9

    return sum(digits) % 10 == 0

File: go/luhn/luhn.go
package luhn

import "unicode"
import "fmt"

// Determines if number is valid per the Luhn formula
func Valid(number string) bool {
	digits, err := normalize(number)

	if err != nil {
		return false
	}

	if len(digits) == 1 {
		return false
	}

	sum := 0

	for i, digit := range digits {
		sum += checksum(i, digit)
	}

	return sum%10 == 0
}

// converts to int, removes spaces and reverses
func normalize(number string) ([]int, error) {
	digits := []rune(number)
	normalizedDigits := []int{}
	var digit rune

	for i, count := 0, len(digits); i < count; i++ {
		digit = digits[count-i-1]

		if unicode.IsSpace(digit) {
			continue
		}

		if !unicode.IsDigit(digit) {
			return []int{}, fmt.Errorf("only digits are allowed. found: %v", digit)
		}

		normalizedDigits = append(normalizedDigits, toInt(digit))
	}

	return normalizedDigits, nil
}

// converts rune digit to integer
func toInt(digit rune) int {
	return int(digit) - '0'
}

// computes the checksum for the digit
func checksum(i int, digit int) int {
	if i%2 == 0 {
		return digit
	}

	digit *= 2

	if digit > 9 {
		digit -= 9
	}

	return digit
}
